Defibrillator refactor and add defib prediction

TokenStyle opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
Refactored defibrillator's code and made it predictive according to issue #28051
I've set `cvar net.fakelagmin 0.4` to fake lags, defibs predictions works great.
<!-- What did you change in this PR? -->

## Why / Balance
More smooth experience using defibrillator overall.
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->

## Technical details
Changed defibs code, make prediction. And my thoughts:
A lot of possibilities to make predictive and refactor to `PowerCellSystem`, gauges doAfter, and med overall, but it would take a lot of work and I'm too lazy to do it.
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->

## Media

https://github.com/user-attachments/assets/44a49300-5ec5-49ff-86d5-dbaa15b8acff


<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
Defibrillator refactor and made it predictive, add defibs code in Content.Shared, Content.Client.
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->
no :cl: no fun
<!--
Make sure to take this Changelog template out of the comment block in order for it to show up.
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->

deltanedas commented on Content.Shared/Medical/DefibrillatorComponent.cs:
```diff
@@ -12,38 +12,38 @@ namespace Content.Shared.Medical;
     /// The time at which the zap cooldown will be completed
     /// </summary>
-    [DataField("nextZapTime", customTypeSerializer: typeof(TimeOffsetSerializer)), ViewVariables(VVAccess.ReadWrite)]
+    [DataField(customTypeSerializer: typeof(TimeOffsetSerializer)), ViewVariables(VVAccess.ReadWrite)]
```
none of these need viewvariables, all datafields are automatically rw

TokenStyle replied:
yep, removed all `ViewVariables(VVAccess.ReadWrite)`

deltanedas reviewed (approved):
lgtm

TokenStyle commented:
fixed merge conflict, passed tests

ShadowCommander commented on Content.Server/Medical/DefibrillatorSystem.cs:
```diff
@@ -1,151 +1,43 @@
-        {
-            Zap(uid, user, user, component);
-            return;
-        }
-
+        // TODO : powercell should be rewritten to shared instead of strictly be on Server side
         if (!_powerCell.TryUseActivatableCharge(uid, user: user))
             return;
```
This method doesn't call the Shared.Zap. You'll need to call `base.Zap` or something here and maybe add a bool return value for if the shared code returns.

TokenStyle replied:
fixed by calling base.Zap in server's overrided Zap

ShadowCommander reviewed (changes requested):
Thanks for the PR.

I tested the functionality and it was not doing the Electrocution animation. This is because the server code is not calling the shared code for Zap.

deltanedas commented on Content.Shared/Medical/DefibrillatorComponent.cs:
```diff
@@ -12,38 +12,38 @@ namespace Content.Shared.Medical;
     /// The time at which the zap cooldown will be completed
     /// </summary>
-    [DataField("nextZapTime", customTypeSerializer: typeof(TimeOffsetSerializer)), ViewVariables(VVAccess.ReadWrite)]
+    [DataField(customTypeSerializer: typeof(TimeOffsetSerializer)), ViewVariables(VVAccess.ReadWrite)]
```
should also have autopausedfield

TokenStyle replied:
Removed VV read write and add AutoPausedField

TokenStyle commented:
> I tested the functionality and it was not doing the Electrocution animation. This is because the server code is not calling the shared code for Zap.

fixed. I moved `ElectrocutionSystem` to server side, cuz method `TryDoElectrocution` have to be called only in server side, and there's nice comment to remind `// only done serverside`

metalgearsloth reviewed (changes requested):
Are any of the fields even networked?

TokenStyle commented:
> Are any of the fields even networked?

I got it. I previously thought it's handled by `AutoGenerateComponentPause` and `AutoPausedField`, but these attributes doesn't handle network.
Added `AutoNetworkedField` to network it. Network automatically replicated and handled by RobustToolkit.
```cs
/// <summary>
/// The time at which the zap cooldown will be completed.
/// Auto-networked field to sync zap time, and fix that player doing zap before it actually cooldowned
/// </summary>
[DataField(customTypeSerializer: typeof(TimeOffsetSerializer)), AutoNetworkedField, AutoPausedField]
public TimeSpan? NextZapTime;
```

TokenStyle commented:
Month has been passed

TokenStyle commented:
https://github.com/user-attachments/assets/5d821894-102b-47ed-b384-0195429cecc3

Upstreamed and made video

FairlySadPanda commented:
Tested, works fine when integrated with latest master.

VerinSenpai commented:
Did somebody say prediction 👺

TokenStyle commented:
well, everything is works and tested locally, ci tests are passed

Princess-Cheeseballs commented:
Are you still working on this?
